src/frontend/Deprecation_analysis.ml

```diff
@@ -0,0 +1,194 @@
+          [ ( emeta.loc
+            , name ^ " is deprecated and will be removed in the future. Use "
+              ^ rename_deprecated deprecated_distributions name
+              ^ " instead." ) ]
+        else if Option.is_some (String.Map.find deprecated_functions name) then
+          [ ( emeta.loc
+            , name ^ " is deprecated and will be removed in the future. Use "
+              ^ rename_deprecated deprecated_functions name
+              ^ " instead." ) ]
+        else if Option.is_some (String.Map.find deprecated_odes name) then
+          [ ( emeta.loc
+            , name ^ " is deprecated and will be removed in the future. Use "
+              ^ rename_deprecated deprecated_odes name
+              ^ " instead. \n\
+                 The new interface is slightly different, see: \n\
+                 https://mc-stan.org/users/documentation/case-studies/convert_odes.html"
+            ) ]
+        else []
+      in
+      acc @ w
+      @ List.concat
+          (List.map l ~f:(fun e ->
+               warn_deprecated_expr deprecated_userdefined [] e ))
+  | FunApp (UserDefined, {name; _}, l) ->
+      let w =
+        let type_ = String.Map.find deprecated_userdefined name in
+        if Option.is_some type_ then
+          [ ( emeta.loc
+            , "Use of the _log suffix in user defined function " ^ name
+              ^ " is deprecated, use "
+              ^ update_suffix name (Option.value_exn type_)
+              ^ " instead." ) ]
+        else []
+      in
+      acc @ w
+      @ List.concat
+          (List.map l ~f:(fun e ->
+               warn_deprecated_expr deprecated_userdefined [] e ))
+  | _ ->
+      fold_expression
+        (warn_deprecated_expr deprecated_userdefined)
+        (fun l _ -> l)
+        acc expr
+
+let warn_deprecated_lval deprecated_userdefined acc l =
+  fold_lval_with
+    (warn_deprecated_expr deprecated_userdefined)
+    (fun x _ -> x)
+    acc l
+
+let rec warn_deprecated_stmt deprecated_userdefined
+    (acc : (Location_span.t * string) list) {stmt; _} :
+    (Location_span.t * string) list =
+  match stmt with
+  | IncrementLogProb e -> warn_deprecated_expr deprecated_userdefined acc e
+  | Assignment {assign_lhs= l; assign_op= ArrowAssign; assign_rhs= e} ->
+      acc
+      @ warn_deprecated_lval deprecated_userdefined [] l
+      @ warn_deprecated_expr deprecated_userdefined [] e
+  | FunDef {body; _} -> warn_deprecated_stmt deprecated_userdefined acc body
+  | _ ->
+      fold_statement
+        (warn_deprecated_expr deprecated_userdefined)
+        (warn_deprecated_stmt deprecated_userdefined)
+        (warn_deprecated_lval deprecated_userdefined)
+        (fun l _ -> l)
+        acc stmt
+
+let analyze_udfs program =
+  program.functionblock |> Option.value ~default:[]
+  |> List.filter_map ~f:find_suffixes
+  |> List.dedup_and_sort ~compare:(fun (x, _) (y, _) -> String.compare x y)
+  |> String.Map.of_alist_exn
```

Comment:
I'm not quite clear on the difference between `analyze_udfs` and `userdef_distributions`, they both seem to be collecting user-defined functions with appropriate suffixes. Are they both needed?